Remove Enzyme

nemanjaglumac opened the pull request:
### Status
READY

### What does this PR accomplish?
- Converts unit tests from Enzyme to React Testing Library
- Converts some of the unit tests to Cypress e2e tests
- Removes Enzyme-specific hacks and mocks
- Removes every last file that was in any way related to/dependent on Enzyme
- Removes `enzyme` library
- Removes `enzyme-adapter-react-15` library
- Closes #13657 (Main list for all affected tests)

As a side-effect, it also adds [@testing-library/user-event](https://github.com/testing-library/user-event) (and its required library `@testing-library/dom`) that is preferred over default `fireEvent()` API in `@testing-library/react`. This was needed for the work on `TokenField.unit.spec.js`
> Most projects have a few use cases for fireEvent, but the majority of the time you should probably use @testing-library/user-event.
https://testing-library.com/docs/dom-testing-library/api-events/#fireevent

### How to test this?
- `yarn test` - all tests should pass
- All tests should pass in CI as well

### Additional notes:
It would be extremely hard to review all 49 files. Therefore, I've tried to keep meaningful changes in small and focused commits that should be easier to review individually.

- `TokenField.unit.spec.js` was already reviewed in #14223
___
- `AdHocQuestionLoader.unit.spec.js`,
- `SavedQuestionLoader.unit.spec.js` and
- `QuestionLoader.unit.spec.js`
were reviewed in #14224
___
- Integration tests were converted to Cypress in #14174

nemanjaglumac committed:
Convert `AccordionList.unit.spec.js`

nemanjaglumac committed:
Convert `LoadingAndErrorWrapper.unit.spec.js`

nemanjaglumac committed:
Convert `ProfileLink.unit.spec.js`

nemanjaglumac committed:
Convert `QuestionAndResultLoader.unit.spec.js`

nemanjaglumac committed:
Convert `Calendar.unit.spec.js`

nemanjaglumac committed:
Convert `DateRangeWidget.unit.spec.js`

nemanjaglumac committed:
Convert `DatePicker.unit.spec.js`

nemanjaglumac committed:
Convert `Overwlorld.unit.spec.js`

nemanjaglumac committed:
Convert `QuestionResultLoader.unit.spec.js`

nemanjaglumac committed:
Convert `LegendVertical.unit.spec.js`

nemanjaglumac committed:
Convert `ChartSettingOrderedColumns.unit.spec.js`

nemanjaglumac committed:
Convert `ChartNestedSettingsSeries.unit.spec.js`

nemanjaglumac committed:
Convert `ObjectDetail.unit.spec.js`

nemanjaglumac committed:
Convert `FilterPopover.unit.spec.js`

nemanjaglumac committed:
Convert `TimeseriesFilterWidget.unit.spec.js`

nemanjaglumac committed:
Convert `PivotByDrill.unit.spec.js`

nemanjaglumac committed:
Convert `FieldValuesWidget.unit.spec.js`

nemanjaglumac committed:
Convert `Logs.unit.spec.js`

nemanjaglumac committed:
Convert `GuiQueryEditor.unit.spec.js`

nemanjaglumac committed:
Convert `SaveQuestionModal.unit.spec.js`

nemanjaglumac committed:
Add todo note to `DashCard.unit.spec.js`

The test was already skipped.

nemanjaglumac committed:
Convert `EntityListLoader.unit.spec.js` [ci skip]

paulrosenzweig committed:
convert QuestionLoader tests (#14224)

nemanjaglumac committed:
Convert `TokenField.unit.spec.js` from Enzyme to RTL [ci skip] (#14223)

* Add `@testing-library/user-event` library

`@testing-library/dom` is required:
https://github.com/testing-library/user-event#installation

* Convert initial batch of tests

* Convert "when updateOnInputChange is provided" describe block

* Skip still unconverted `describe` blocks

* Convert "when updateOnInputBlur is true" describe block

* Convert "when updateOnInputBlur is false" describe block

* Convert "with multi=true" describe block

* Convert "with multi=false" describe block

* Convert "custom layoutRenderer" describe block

* Delete obsolete code and fix linter errors

* Extract logic for `type()` function

* Rename test

* Extract logic for asserting on the text content for values and options

* Extract logic for clicking on text elements

* Extract logic for entering keyboard keys into input [ci skip]

nemanjaglumac committed:
Fix `DateRangeWidget` test

nemanjaglumac committed:
Convert `pulse.unit.spec.js`

nemanjaglumac committed:
Convert `Background.unit.spec.js`

nemanjaglumac committed:
Convert `LoginApp.unit.spec.js` to Cypress tests

There are "normal" and enterprise version of these tests. This commit covers both.

nemanjaglumac committed:
Delete `integration.js` that is now no longer in use

nemanjaglumac committed:
Delete `TimeseriesMode.unit.spec.js`

It was introduced originally because `TimeseriesGroupingWidget` component was commented out.
https://github.com/metabase/metabase/commit/77877896b2fda41455b9f459fad4956cee1578b1

nemanjaglumac committed:
Remove Enzyme

nemanjaglumac committed:
Remove Enzyme-specific "hacks" and mocks

nemanjaglumac committed:
Fix linter complaining about undefined value in skipped test

nemanjaglumac commented on frontend/test/__support__/mocks.js:
```diff
@@ -34,21 +34,8 @@ jest.mock("ace/snippets/json", () => {}, { virtual: true });
 jest.mock("ace/snippets/json", () => {}, { virtual: true });
 jest.mock("ace/ext-language_tools", () => {}, { virtual: true });
 
-// Use test versions of components that are normally rendered to document root or use unsupported browser APIs
-import * as modal from "metabase/components/Modal";
-modal.default = modal.TestModal;
-
-import * as tooltip from "metabase/components/Tooltip";
-tooltip.default = tooltip.TestTooltip;
-
 jest.mock("metabase/components/Popover");
 
-import * as bodyComponent from "metabase/components/BodyComponent";
-bodyComponent.default = bodyComponent.TestBodyComponent;
-
-import * as table from "metabase/visualizations/visualizations/Table";
-table.default = table.TestTable;
-
```
I wasn't sure if any of these mocks were actually used anywhere, but:
1. I couldn't find any reference
2. All tests are passing without them

If I've overlooked something, please let me know and I can easily revert this commit. It's not crucial for Enzyme removal.

kdoh commented on frontend/test/metabase/pulse/pulse.unit.spec.js:
```diff
@@ -1,57 +1,46 @@
-  { id: 1, common_name: "Barb", email: "barb_holland@hawkins.mail" }, // w
-  { id: 2, common_name: "Dustin", email: "dustin_henderson@hawkinsav.club" }, // w
-  { id: 3, common_name: "El", email: "011@energy.gov" },
-  { id: 4, common_name: "Lucas", email: "lucas.sinclair@hawkins.mail" }, // w
```
Any reason to change these? While these were originally just a fun Stranger Things joke, I do wonder if having different TLDs is a nice thing to keep around in case it were to expose something weird in a email regex or some such. (Although I realize that's not being tested here).

nemanjaglumac replied:
@flamber argued many times before that we shouldn't use real TLDs. [`.test`](https://en.wikipedia.org/wiki/.test) is reserved TLD intended for exactly these kind of purposes. I started using it exclusively in all the tests ever since I first learned about reserved domains.

kdoh replied:
Fair enough.

kdoh commented on frontend/test/metabase/dashboard/DashCard.unit.spec.js:
```diff
@@ -4,7 +4,9 @@ jest.unmock("metabase/components/ExplicitSize");
 import React from "react";
 import renderer from "react-test-renderer";
-import { render } from "enzyme";
+// TODO: re-write using React Testing Library (Enzyme is removed!)
```
Can this be removed?

nemanjaglumac replied:
I added some comments to each individual test. If we can find a way to test slow loading cards in Cypress without affecting the total testing time drastically, this whole test should be removed, imo.

Can we add hard coded slow loading card to our internal library and make one specific Cypress test for that like @paulrosenzweig did recently?

nemanjaglumac replied:
I removed unused parts in https://github.com/metabase/metabase/pull/14094/commits/c0f78ddb993cfd8e8c85dfeb5cafa3952e53374f.

The only one left is `.Card--slow`. I'm on board with leaving this in a form of a unit test. I didn't know how to make it work with RTL (requires a `Provider` and a `store`), but if you want to give it a try, please go for it.

paulrosenzweig replied:
Oh yeah I forgot that DashCard is connected. I think that was my fault. It should be possible for a small refactor to enable unit testing, but that can happen later.

kdoh commented on frontend/test/metabase/containers/Overworld.unit.spec.js:
```diff
@@ -1,38 +1,57 @@
+import "@testing-library/jest-dom/extend-expect";
+import { fireEvent, render, screen } from "@testing-library/react";
+
+const STRING_1 = "Your team's most important dashboards go here";
```
I think I'd prefer slightly more descriptive const names here.

kdoh commented on frontend/src/metabase/containers/QuestionLoader.jsx:
```diff
@@ -71,7 +71,7 @@ const QuestionLoader = ({
     />
   ) : // if there's a questionHash it means we're in ad-hoc land
-  questionHash != null ? (
+  questionHash != null && questionHash !== "" ? (
```
What's the reason for the additional check?

nemanjaglumac replied:
This was a code from the previous PR. Paul explained it here:
https://github.com/metabase/metabase/pull/14224#discussion_r549896224

kdoh replied:
Ah got it. Thanks.

kdoh reviewed (commented):
Still working my way through things, (about halfway through) but so far looking really solid. Just a few minor notes at this point.

nemanjaglumac committed:
Use more descriptive const names

nemanjaglumac committed:
Tidy up comments in `DashCard.unit.spec.js` [ci skip]

paulrosenzweig commented on frontend/test/metabase/visualizations/components/settings/ChartNestedSettingsSeries.unit.spec.js:
```diff
@@ -17,28 +18,28 @@ function getSeries(display) {
-    expect(settings.find(".Icon-line")).toHaveLength(1);
-    expect(settings.find(".Icon-area")).toHaveLength(1);
-    expect(settings.find(".Icon-bar")).toHaveLength(1);
+    getByRole("img", { name: /line/i });
```
Probably requires a RTL plugin, but it could be nice to have `*ByIconName("line")`

nemanjaglumac replied:
We can probably make a helper function for that.

paulrosenzweig commented on frontend/test/metabase/visualizations/components/settings/ChartSettingOrderedColumns.unit.spec.js:
```diff
@@ -18,61 +16,70 @@ function renderChartSettingOrderedColumns(props) {
-      [[{ name: "Bar", enabled: true }, { name: "Foo", enabled: true }]],
-    ]);
+
+  // TODO: Test this in Cypress
```
👍 on moving this to Cypress. That both allows a more real drag-and-drop, but also reordering columns requires a re-run to actually display the new table.

paulrosenzweig commented on package.json:
```diff
@@ -118,8 +120,6 @@
     "css-loader": "^0.28.7",
     "cypress": "^5.5.0",
     "documentation": "^4.0.0-rc.1",
-    "enzyme": "^3.8.0",
```
🎉

paulrosenzweig reviewed (approved):
Awesome job pushing through all of these!

nemanjaglumac commented on enterprise/frontend/test/metabase-enterprise/auth/sso.cy.spec.js:
```diff
@@ -0,0 +1,28 @@
+  describeWithToken,
+} from "__support__/cypress";
+
+describeWithToken("scenarios > auth > signin > SSO", () => {
```
There is non-enterprise version of this test as well. If I understood correctly, the only difference between EE and OSS versions is that EE admin is able to completely revoke logging in with password (which was a fallback login option in OSS). This test tests exactly that scenario.

paulrosenzweig replied:
Oops yes that's right! The `enable-password-login` part is EE only.

nemanjaglumac commented on frontend/test/metabase/components/Calendar.unit.spec.js:
```diff
@@ -22,14 +22,16 @@ describe("Calendar", () => {
 
   it("should switch months correctly", () => {
```
Thank you.

re: readability - I totally agree. Very often writing unit tests with RTL feels like writing e2e with Cypress.

nemanjaglumac committed:
Remove obsolete tests from `DashCard.unit.spec.js` [ci skip]

- `.Card--recent` has been removed in https://github.com/metabase/metabase/pull/14269
- `.Card--unmapped` was removed prior to these changes, but I didn't notice it wasn't used until now.

nemanjaglumac commented on frontend/test/metabase/components/LoadingAndErrorWrapper.unit.spec.js:
```diff
@@ -1,43 +1,42 @@
+      render(<LoadingAndErrorWrapper loading={false} noWrapper />);
     });
 
     it("should display a given scene during loading", () => {
```
Maybe this was just a fail-safe mechanism that somebody wanted to test? Would you like me to remove it or leave it as it is?

nemanjaglumac commented on frontend/test/metabase/containers/QuestionAndResultLoader.unit.spec.js:
```diff
@@ -1,10 +1,10 @@
 describe("QuestionAndResultLoader", () => {
   it("should load a question and a result", () => {
-    shallow(<QuestionAndResultLoader>{() => <div />}</QuestionAndResultLoader>);
+    render(<QuestionAndResultLoader>{() => <div />}</QuestionAndResultLoader>);
```
Yeah. It feels like it was just a setup for some future test, but it was left at that.
